llvm/lib/Transforms/IPO/MergeFunctions.cpp

```diff
@@ -375,9 +375,35 @@ bool MergeFunctions::doFunctionalCheck(std::vector<WeakTrackingVH> &Worklist) {
 }
 #endif
 
+/// Check whether \p F has an intrinsic which references
+/// distinct metadata as an operand. The most common
+/// instance of this would be CFI checks for function-local types.
+static bool hasDistinctMetadataIntrinsic(const Function &F) {
+  for (const BasicBlock &BB : F) {
+    for (const Instruction &I : BB.instructionsWithoutDebug()) {
+      if (!isa<IntrinsicInst>(&I))
+        continue;
+
+      for (unsigned i = 0, e = I.getNumOperands(); i != e; ++i) {
+        MetadataAsValue *MDL = dyn_cast<MetadataAsValue>(I.getOperand(i));
+        if (!MDL)
+          continue;
+        MDNode *node = dyn_cast<MDNode>(MDL->getMetadata());
+        if (!node)
+          continue;
+        if (!node->isDistinct())
+          continue;
```

Comment:
Since there isn't branching and the deepest block is short, it should be fine to nest here:
```
if (MDNode *N = ...)
  if (N->isDistinct())
    return true;
```